### DISCUSSION TITLE
Status Update: 2022-05-11

### DISCUSSION THREAD

**Discussion by haixuanTao** [dora-rs/dora#25] [Status Update]
In this discussion, we describe our recent progress and the main topics that we would like to discuss in today's meeting. We use a separate post for each topic to keep the discussion focused. Feel free to create new posts for topics that are not mentioned yet.

**Comment by haixuanTao** [comment]
# Meeting 2022-05-11

# Messages Design

Metadata field in messages help us keep track of informations throughout the distributed system, such as:
- Watermark
- Deadline
- Opentelemetry (spans and log)

#### Exemple with Capnproto

```capnproto
struct Metadata {
  metadataVersion @0 :Text;
  watermark @1 :UInt64; 
  deadline @2 :UInt64;
  openTelemetryContext @3 :Text;
  depth @4 :UInt32; # Depth within the graph of the message 
}


struct Message {
  id @0 :UInt32;
  data @1 :Data; # ByteArray
  metadata @2 :Metadata;
}

```

**Reply by haixuanTao** [reply]
Remark: 
Metadataversion -> Text -> UInt32
Metadata @2 -> Metadata @1

**Reply by haixuanTao** [reply]
Future design ideas:
- Make Data Format configurable.

**Reply by haixuanTao** [reply]
Older Serialization benchmark: [https://blog.logrocket.com/rust-serialization-whats-ready-for-production-today/](https://blog.logrocket.com/rust-serialization-whats-ready-for-production-today/)
New kid on the block: protocol: [https://github.com/otake84/dlhn](https://github.com/otake84/dlhn)

**Comment by haixuanTao** [comment]
## OpenTelemetry with Jaeger
![](https://i.imgur.com/Dqut5oC.png)

- Cross language telemetry.
- Distributed telemetry for dora-rs, enabling tracing for the whole graph.
- Records spans and logs across language.
- Queryable metrics enabling academic research but also large scale production system.
- Can be integrated to Influxdb, Prometheus, Datadog.

**Reply by haixuanTao** [reply]
[https://github.com/tokio-rs/tracing](https://st1.zoom.us/web_client/1q1nf58/html/externalLinkPage.html?ref=https://github.com/tokio-rs/tracing)

**Comment by haixuanTao** [comment]
## Implementation progress

- Improvement on shared libraries C-API.
- Progress on nodes and operators implementation.
- Progress on the data plane and control plane part of the middleware.

**Comment by heyong4725** [comment]
Zenoh team may update their API further to make it better, here is the discussion: https://github.com/eclipse-zenoh/roadmap/discussions/23
Also there are good discussion on Rust async/sync: https://github.com/eclipse-zenoh/roadmap/discussions/17
Hope this is useful for our Dora work

**Comment by heyong4725** [comment]
Also Zenoh Flow is refactoring new deadline implementation: https://github.com/eclipse-zenoh/roadmap/discussions/22
@haixuanTao @phil-opp

**Comment by heyong4725** [comment]
@phil-opp you may want to take a look of this repo: https://github.com/tower-rs/tower/tree/master/tower-service 
Tower-service is trying to abstract the request/response comm pattern called service, which is common, and similar to ROS2 service abstraction. It seems that tower use http as transport, but you could potentially use other transport.
It may be part of our middleware abstraction as well, and potentially for essa project as well.
